### ISSUE TITLE
Preparing the data for pRoloc

### ISSUE THREAD

**Issue by lgatto** [lgatto/pRoloc#56]
Here is a question regarding data preparation I got by email:

> I would like to use your pRoloc package for supervised classification. I am finding difficulty in understanding on the how to make my dataset according to the pRoloc standard dataset.
> 
> For example,
> 
> I have a matrix *A* of size m, n (where m is number of samples, and n is number of features) and I have vector *label* of size mx1 (which indicates the class label of each sample). Could you please tell me, how to make into pRoloc data format or how to use as such *A*, and *label*

**Comment by lgatto** [comment]
Let's simulate some data to get started. 

```r
> A <- matrix(rnorm(180), ncol = 9, nrow = 20)
> colnames(A) <- paste0("sample", 1:9)
> rownames(A) <- paste0("feature", 1:20)
> head(A)
            sample1    sample2    sample3    sample4    sample5    sample6
feature1  0.6058378  0.2021587 -1.1590380 -0.4684858 -0.2219107  0.9169211
feature2 -0.7426731 -0.9723679 -0.1144474  0.5328933 -1.7608863 -0.8793810
feature3  0.8654833 -0.5335896 -2.2735659 -1.7175923  0.5879435  0.6963291
feature4  0.1624961  0.3472060  0.3556253  0.4966961 -0.4786782  1.6067455
feature5 -1.0672379 -0.9184442  1.5294448  0.2919177 -1.6221086 -1.3069077
feature6  0.6690946 -0.6117331 -0.3890310 -0.9516129  1.4275854  0.6593734
            sample7    sample8     sample9
feature1  0.5238955 -0.0446173 -0.90228151
feature2  0.2798391  0.3190033  1.36583980
feature3  0.2263847  0.9368059 -0.09086595
feature4 -1.0148659 -0.6861671  0.81093210
feature5 -0.3571901 -1.7188145 -0.38704546
feature6 -0.1606454 -0.2228701 -1.92473667
```
The question states that the sample are labelled. Typically, in `pRoloc`, we label features, but that should not be a problem. 
```r
> labels <- rep(c("A", "B", "unknown"), 3)
> labels
[1] "A"       "B"       "unknown" "A"       "B"       "unknown" "A"      
[8] "B"       "unknown"
```
We now need to generate the `MSnSet`, the basic data structure used in all `pRoloc`. Typically, the features are along the rows and samples along the columns, so let's do that. We need to defined feature and sample metadata, named `fd` (feature data) and `pd` (pheno data) below. These are simple dataframes.

```r

> pd <- data.frame(markers = labels, row.names = colnames(A))
> pd
        markers
sample1       A
sample2       B
sample3 unknown
sample4       A
sample5       B
sample6 unknown
sample7       A
sample8       B
sample9 unknown
> fd <- data.frame(id <- 1:20, row.names = rownames(A))
> head(fd)
         id....1.20
feature1          1
feature2          2
feature3          3
feature4          4
feature5          5
feature6          6
```
Of course, these would typically contain important information, such feature annotation, some relevant feature quality scores, .... In a typical `pRoloc` data, the `markers` would be stored as feature metadata.

We can now create the `MSnSet`:

```r
> x <- MSnSet(exprs =A, fData = fd, pData = pd)
> x
MSnSet (storageMode: lockedEnvironment)
assayData: 20 features, 9 samples 
  element names: exprs 
protocolData: none
phenoData
  sampleNames: sample1 sample2 ... sample9 (9 total)
  varLabels: markers
  varMetadata: labelDescription
featureData
  featureNames: feature1 feature2 ... feature20 (20 total)
  fvarLabels: id....1.20
  fvarMetadata: labelDescription
experimentData: use 'experimentData(object)'
Annotation:  
- - - Processing information - - -
 MSnbase version: 1.17.3 
```

But, the question pointed out that the samples are the items to be classified. We thus need to transpose the `MSnSet` before being able to use the `pRoloc` functions out of the box.

```r
> tx <- t(x)
> tx
MSnSet (storageMode: lockedEnvironment)
assayData: 9 features, 20 samples 
  element names: exprs 
protocolData: none
phenoData
  sampleNames: feature1 feature2 ... feature20 (20 total)
  varLabels: id....1.20
  varMetadata: labelDescription
featureData
  featureNames: sample1 sample2 ... sample9 (9 total)
  fvarLabels: markers
  fvarMetadata: labelDescription
experimentData: use 'experimentData(object)'
Annotation:  
- - - Processing information - - -
MSnSet transposed: Thu May 28 13:07:39 2015 
 MSnbase version: 1.17.3 
> fData(tx)
        markers
sample1       A
sample2       B
sample3 unknown
sample4       A
sample5       B
sample6 unknown
sample7       A
sample8       B
sample9 unknown
```

**Comment by bharathbhushandamodaran** [comment]
Thank for very much for your comments. Thanks I am able to generate a MSnSet datatypes to be used in pRoloc package. However, I have came across another problem while using (for e.g : tx in classification).
when I give "tx" for the optimizing the classifiers, it works fine. But, when I give same "tx" for training the classification, I get error.

params <- perTurboOptimisation(tx,
                                pRegul = 2^seq(-2,2,2),
                                sigma = 10^seq(-1, 1, 1),
                                inv = "Inversion Cholesky",
                              reg ="tikhonov",
                              times = 3)
this works well.. but in the next step
res <- perTurboClassification(tx, params)
I get the error as below:
Error in res[, i] : subscript out of bounds.

similarly, I tried to with other classifier, for e.g SVM
params <- svmOptimisation(tx, cost = 2^seq(-2,2,2), sigma = 10^seq(-1, 1, 1), times = 3)
it works well, but when I train the SVM classifier, 

 res <- svmClassification(tx, params),  I have the following error
Error in matrix(ret$dec, nrow = nrow(newdata), byrow = TRUE, dimnames = list(rowns,  : 
  length of 'dimnames' [2] not equal to array extent.

similarly, when I performed nnet 

params <- nnetOptimisation(dunkley2006, decay = 10^(c(-1, -5)), size = c(5, 10), times = 3)
 it works well, but training had problem...
res <- nnetClassification(tx, params)
# weights:  41
initial  value 107.193063 
iter  10 value 50.003043
final  value 50.000653 
converged
Error in predict.nnet(obj, teData, type = "class") : 
  inappropriate fit for class 

Could you please tell me how to resolve it..

**Comment by lgatto** [comment]
Could you plot your data as follows
```r
plot2D(tx, fcol = "markers"
```
and add show the result, and then we will see from there.

**Comment by bharathbhushandamodaran** [comment]
Thanks for your response. Please see below for the plot of my data...

![rplot](https://cloud.githubusercontent.com/assets/12643715/7868120/31279344-057a-11e5-8905-fb3c8bf0ed5f.png)

**Comment by lgatto** [comment]
From your plot, it looks like there are not unlabelled row features to be classified. They are, by convention, labelled as `"unknown"` in `pRoloc`.

**Comment by bharathbhushandamodaran** [comment]
Thanks for your comment. Yes, I know the labels for all of my row features (i.e samples in convention machine learning framework) and more over this is a training samples (features). 

I have separate samples set (testing set or validation set), used for testing the classifier (or to predict the class by the algorithm). Do do mean "unkown" samples are as the testing samples (samples/features to be classified after training stage of the classifier).

**Comment by lgatto** [comment]
> I have separate samples set (testing set or validation set), used for testing the classifier (or to predict the class by the algorithm). Do do mean "unkown" samples are as the testing samples (samples/features to be classified after training stage of the classifier).

Yes, the convention is to combine all the data [1] - labelled, training, testing and validation - into one `MSnSet` (`tx` above) and the sample whose labels need to be inferred are defined using `"unknown"`. This data is automatically split into `"unknwon"`'s and other labels. These *other* labels are used to optimise hyper-parameters (using the `___Optimisation` functions, that will perform stratified splitting and cross-validation [2]). This result (called a `GenRegRes` object) should then be visually checked. Then, the same data and hyper-parameters from the previous step are passed to the corresponding `___Classification` function and only the sample labelled as `"unknown"` are classified. All these steps are described in great detail, with reproducible examples, in the `pRoloc-tutorial` vignette.

[1] This enable to pre-process (normalisation, missing data imputation, ...) and visualise the complete data, before and after each step of the analysis, including classification. 
[2] `___` being one of `svm`, `perTurbo`, `nnet`, `knn`, ...

**Comment by bharathbhushandamodaran** [comment]
Thanks very much for your answers and also for he reference materials. Now it is working fine.

**Comment by lgatto** [comment]
Great, I will close this issue, then. Please open a new one if you have another question.
